hail/python/hail/plot/plots.py

```diff
@@ -126,6 +127,147 @@ def cumulative_histogram(data, range=None, bins=50, legend=None, title=None, nor
+
+    If x_range or y_range are not provided, the function will do a pass through the data to determine min and max
+    of each variable.
+
+    Parameters
+    ----------
+    x : :class:`.NumericExpression`
+        Expression for x-axis.
+    y : :class:`.NumericExpression`
+        Expression for y-axis.
+    x_range : Tuple[float]
+        Tuple of (min, max) bounds for the x-axis.
+    y_range : Tuple[float]
+        Tuple of (min, max) bounds for the y-axis.
+    bins : int
+        Number of bins in the histogram (default 40).
+    x_bins : int
+        Number of bins on x-axis, will override ``bins`` if provided.
+    y_bins : int
+        Number of bins on y-axis, will override ``bins`` if provided.
+    plot_width : int
+        Plot width (default 600px).
+    plot_height : int
+        Plot height (default 600px).
+    plot_title : str
+        Title of the plot.
+    font_size : str
+        String of font size in points (default "7pt").
+    colors : List[str]
+        List of hex colors from low to high.
+
+    Returns
+    -------
+    :class:`bokeh.plotting.figure.Figure`
+    """
+    source = x._indices.source
+    y_source = y._indices.source
+
+    if source is None or y_source is None:
+        raise ValueError("histogram_2d expects two expressions of 'Table', found scalar expression")
+    if isinstance(source, hail.MatrixTable):
+        raise ValueError("hisogram_2d requires source to be Table, not MatrixTable")
+    check_row_indexed('histogram_2d', x)
+    check_row_indexed('histogram_2d', y)
+    if source != y_source:
+        raise ValueError(f"histogram_2d expects two expressions of 'Table', found {source} and {y_source}")
+    if x_bins is None:
+        x_bins = bins
+    if y_bins is None:
+        y_bins = bins
+    if x_range is None or y_range is None:
+        warnings.warn('At least one range was not defined. Doing two passes...')
+        ranges = source.aggregate(hail.struct(x_stats=hail.agg.stats(x),
+                                              y_stats=hail.agg.stats(y)))
+        if x_range is None:
+            x_range = (ranges.x_stats.min, ranges.x_stats.max)
+        if y_range is None:
+            y_range = (ranges.y_stats.min, ranges.y_stats.max)
+    x_range = list(map(float, x_range))
+    y_range = list(map(float, y_range))
+    x_spacing = (x_range[1] - x_range[0]) / x_bins
+    y_spacing = (y_range[1] - y_range[0]) / y_bins
+
+    def frange(start, stop, step):
+        from itertools import count, takewhile
+        return takewhile(lambda x: x < stop, count(start, step))
+
+    x_levels = hail.literal(list(frange(x_range[0], x_range[1], x_spacing))[::-1])
+    y_levels = hail.literal(list(frange(y_range[0], y_range[1], y_spacing))[::-1])
+
+    grouped_ht = source.group_by(
+        x=hail.str(x_levels.find(lambda w: x >= w)), y=hail.str(y_levels.find(lambda w: y >= w))
+    ).aggregate(c=hail.agg.count())
```

Comment:
is this the right indentation?

Reply:
My PyCharm claims it is but happy to change if there's another preferred one